audio/out/ao_aaudio.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * AAudio audio output driver.
+ * Copyright (C) 2016 Ilya Zhuravlev <whatever@xyz.is>
+ *
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ao.h"
+#include "internal.h"
+#include "common/msg.h"
+#include "audio/format.h"
+#include "options/m_option.h"
+#include "osdep/timer.h"
+
+#include <aaudio/AAudio.h>
+
+struct priv {
+    AAudioStreamBuilder *builder;
+    AAudioStream *stream;
+    aaudio_performance_mode_t perf_mode;
+};
+
+static void uninit(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    AAudioStream_close(p->stream);
+    AAudioStreamBuilder_delete(p->builder);
+}
+
+static aaudio_data_callback_result_t buffer_callback(AAudioStream *stream, void *userData, void *audioData, int32_t numFrames) {
+    struct ao *ao = userData;
+    double delay;
+
+    //Copied from ao_opensles; what's the 2?
+    delay = 2 * numFrames / (double)ao->samplerate;
+    ao_read_data(ao, &audioData, numFrames, mp_time_us() + 1000000LL * delay);
+
+    //Always?
```

Comment:
Always returning success is fine. This callback will only be invoked while in a playing state.